crates/kitsune_p2p/types/src/dht_arc.rs

```diff
@@ -172,147 +389,259 @@ impl RangeBounds<u32> for ArcRange {
 
-        assert!(DhtArc::new(0, 0).range().is_empty());
-        assert_eq!(DhtArc::new(0, 1).range().into_inc(), 0..=0);
-        assert_eq!(DhtArc::new(1, 2).range().into_inc(), 0..=2);
-        assert_eq!(
-            DhtArc::new(quarter, quarter + 1).range().into_inc(),
-            0..=(half as usize)
-        );
-        check_bounds(quarter, quarter + 1, 0, half);
-
-        assert_eq!(
-            DhtArc::new(half, quarter + 1).range().into_inc(),
-            (quarter as usize)..=((quarter * 3) as usize)
-        );
-        check_bounds(half, quarter + 1, quarter, quarter * 3);
-
-        assert_eq!(
-            DhtArc::new(half, MAX_HALF_LENGTH).range().into_inc(),
-            0..=(u32::MAX as usize)
-        );
-        check_bounds_full(half, MAX_HALF_LENGTH, 0, u32::MAX);
-
-        assert_eq!(
-            DhtArc::new(half, MAX_HALF_LENGTH - 2).range().into_inc(),
-            2..=((u32::MAX - 1) as usize)
-        );
-        check_bounds(half, MAX_HALF_LENGTH - 2, 2, u32::MAX - 1);
-
-        assert_eq!(
-            DhtArc::new(0, 2).range(),
-            ArcRange {
-                start: Included(u32::MAX),
-                end: Included(1)
+    /// Get the density of this bucket.
+    pub fn density(&self) -> PeerDensity {
+        self.arcs
+            .iter()
+            .fold(PeerDensity::new(self.filter, 0.0, 0), |mut d, arc| {
+                d.average_coverage =
+                    (d.average_coverage * d.count as f64 + arc.coverage()) / (d.count as f64 + 1.0);
+                d.count += 1;
+                d
+            })
+    }
+}
+
+impl std::fmt::Display for DhtArcBucket {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        for a in &self.arcs {
+            writeln!(f, "{}", a)?;
+        }
+        writeln!(f, "{} <- Bucket arc", self.filter)
+    }
+}
+
+impl std::fmt::Display for DhtArc {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut out = ["_"; 100];
+        let half_cov = (self.coverage() * 50.0) as isize;
+        let center = self.center_loc.0 .0 as f64 / U32_LEN as f64;
+        let center = (center * 100.0) as isize;
+        for mut i in (center - half_cov)..(center + half_cov) {
+            if i >= 100 {
+                i -= 100;
+            }
+            if i < 0 {
+                i += 100;
             }
-        );
-        check_bounds(0, 2, u32::MAX, 1);
+            out[i as usize] = "#";
+        }
+        out[center as usize] = "|";
+        let out: String = out.iter().map(|a| a.chars()).flatten().collect();
+        writeln!(f, "[{}]", out)
+    }
+}
 
-        assert_eq!(
-            DhtArc::new(u32::MAX, 2).range(),
-            ArcRange {
-                start: Included(u32::MAX - 1),
-                end: Included(0)
+/// Check a set of peers for a gap in coverage.
+/// Note this function is only used for verification in tests at this time.
+pub fn check_for_gaps(peers: Vec<DhtArc>) -> bool {
```

Comment:
```suggestion
#[cfg(test)]
pub fn test_check_for_gaps(peers: Vec<DhtArc>) -> bool {
```

I don't want us to fall into the trap of thinking we have a consistent view of the system and start making production logic choices based on something like this. I vote for making it more explicitly just test logic : )

Reply:
Problem is I need to use it in another crate otherwise I totally would make it test only. I could add a test_utils feature flag to kitsune types if you recon it's worth it?

Reply:
yeah, i think it's worth it : )

Reply:
These two test functions are really long, it would be nice to just move them to a different mod so this one can be for the actual DhtArc definition